### PR TITLE
Improve handling of save-only submissions in stats

### PR DESCRIPTION
**Pull request by jonatanschroeder**
As discussed on Slack. The current handling of question stats does not handle save-only submissions appropriately. This PR fixes the following issues:

* First and last submission score only includes save-and-grade submissions;
* Score arrays for score/point increments use NULL for positions corresponding to save-only submissions;
* Fixes the calculation of point increments by skipping save-only submissions in the array product.

### CODE REVIEW

File: apps/prairielearn/src/sprocs/instance_questions_calculate_stats.sql [hunk 1]
```diff
@@ -9,6 +9,7 @@ WITH first_calculation AS (
         coalesce(bool_or(s.score = 1), FALSE)  AS some_perfect_submission_var,
         coalesce(bool_or(s.score != 0), FALSE) AS some_nonzero_submission_var,
         array_agg(s.score ORDER BY s.date)     AS submission_score_array_var,
+        array_agg(s.score ORDER BY s.date) FILTER (WHERE s.score IS NOT NULL) AS submission_non_null_score_array_var,
```
Comment (nwalters512) [inline apps/prairielearn/src/sprocs/instance_questions_calculate_stats.sql:12]: As long as we're here, might as well get rid of the bespoke formatting on the other lines that tried to vertically align the aliases? Especially now that this line doesn't match the others.

**Review by nwalters512 — COMMENTED** [review]
Before approving, I want to consider whether or not we should do something so that existing data will be (or at least can be) updated to account for this change. Some options:

- Do nothing; any instance question with a submission in the future will be updated when it's scored.
- Run a batched migration to apply this to all existing instance questions and assessment questions.
- Run a batched migration to only apply this to recently-modified instance questions (say, anything modified in 2023).
- Give instructors a "recompute statistics" button.

I suppose I lean towards the first option, just because it's easiest.

File: apps/prairielearn/src/sprocs/instance_questions_calculate_stats.sql (hunk 1 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/sprocs/instance_questions_calculate_stats.sql:12]: Done in 4c0fc0f7b73fec8205f6d5a89e43939fe0aea597.

**Comment by jonatanschroeder** [comment]
From a dev meeting:

* [ ] Create a test for this behaviour.

**Comment by jonatanschroeder** [comment]
@nwalters512 I'm working on the test for this, and the issue now is that `instance_questions_calculate_stats` is called on `grade`, not on `save`. This means that, in a sequence like save-grade-grade-save, the stats will have three entries in the submission array (the first save and two grades), while a following grade would update these to five entries. In other words, if the last submission is a save-without-grade, the stats will not include that last submission until it is either graded or a new grade submission is created. Should this behaviour remain?

**Comment by nwalters512** [comment]
The "stats" in this situation are just the `NULL` entries in our various arrays, correct? I think it would make sense to run `instance_questions_calculate_stats` on all submissions, not just graded ones, so the arrays behave consistently in all cases. I'd like to get Matt's opinion too though.

**Comment by jonatanschroeder** [comment]
> The "stats" in this situation are just the `NULL` entries in our various arrays, correct? I think it would make sense to run `instance_questions_calculate_stats` on all submissions, not just graded ones, so the arrays behave consistently in all cases. I'd like to get Matt's opinion too though.

To make it clear, with an example: say a student saves a submission, then grades the second one with 60%, then saves a third submission. At this point `submission_score_array` would be `[null, 0.6]`, with the third submission not included. If after this point they grade a fourth submission with 80%, the array becomes `[null, 0.6, null, 0.8]`.

In essence, the issue is that right now `instance_questions_calculate_stats` is called from `instance_questions_grade`, which is never called for save-only submissions. A way to fix this would be to call this both on grade (as above) and from `submissions_insert`.

**Comment by jonatanschroeder** [comment]
And to clarify: this only affects the arrays (`submission_score_array`, `incremental_submission_score_array`, `incremental_submission_points_array`).

**Comment by nwalters512** [comment]
Gotcha, thanks for the concrete example. I'm in favor of making the change as you described so that instructors can rely on consistent behavior of these arrays.

**Comment by jonatanschroeder** [comment]
> Gotcha, thanks for the concrete example. I'm in favor of making the change as you described so that instructors can rely on consistent behavior of these arrays.

Done in 882c7b746.

**Comment by mwest1066** [comment]
The Slack discussion referenced in the PR description is https://prairielearn.slack.com/archives/C266KEH9A/p1697644879714059

File: apps/prairielearn/src/sprocs/array_product.sql [hunk 2]
```diff
@@ -5,7 +5,17 @@ CREATE FUNCTION
...
 BEGIN
-    SELECT array_agg(xi * yi) INTO product FROM unnest(x, y) as tmp (xi,yi);
+    FOR i in 1..coalesce(cardinality(x), 0) LOOP
+        IF x[i] IS NULL THEN
+            product[i] := NULL;
+        ELSE
+            product[i] := x[i] * y[j];
+            j := j + 1;
+        END IF;
+    END LOOP;
```
Comment (mwest1066) [inline apps/prairielearn/src/sprocs/array_product.sql:19]: I'm on board with this change but can we change the name of this function to something like `scores_to_points()`? It's no longer really doing a simple `array_product`. It's only used in one place so renaming should be safe.

**Review by mwest1066 — APPROVED** [review]

File: apps/prairielearn/src/sprocs/array_product.sql (hunk 2 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/sprocs/array_product.sql]: Done. Also renamed the parameters for clarity.

Comment (mwest1066) [reply apps/prairielearn/src/sprocs/array_product.sql]: Looks good!

**Pull request merged**